### ISSUE TITLE
Terraform v0.12.7  Crashes when I try to created object list variable

### ISSUE THREAD

**Issue by jwyou1991** [hashicorp/terraform#23508]
Try to create object list variable for the AWS IAM customized policy. Run validate with the variable tf and it crashes. Below is the variable definition:
```
variable "custom"{
    type = list(object({
        action = any
        resource = any
        affect = string
        condition = any
    }))
    default = [
        {
            action = ["s3:Get*",
                        "s3:List*",
                    ]
            resource = ["arn:aws:s3:::mytest1",
                    "arn:aws:s3:::mytest2",
                    ]
            affect = "Allow"
            condition = {}
        },
        {
            action = ["s3:CreateBucket",
                ]
            resource = ["arn:aws:s3:::mybucket-test3",
                ]
            affect = "Allow"
            condition = {}
        }
    ]
}
```

**Comment by jwyou1991** [comment]
Additional note on the variable setup, originally I would like to set the object.action & object.resource to be list type, but terraform would complain about invalid type specification 

> Error: Invalid type specification

  on variable.tf line 3, in variable "custom":
   3:         action = list

The list type constructor requires one argument specifying the element type.

so I change it to type any

**Comment by jwyou1991** [comment]
Let me know if you need the crash log or not for further investigation.

**Comment by danieldreier** [comment]
Thanks for the report! Have you been able to reproduce this on 0.12.16? And yes, please do include the crash log.

**Comment by jwyou1991** [comment]
Hi, @danieldreier . Yes, I am able to reproduce it in 0.12.16 as well. Below is the crash.log

```
2019/11/27 15:47:45 [INFO] Terraform version: 0.12.16
2019/11/27 15:47:45 [INFO] Go runtime version: go1.12.13
2019/11/27 15:47:45 [INFO] CLI args: []string{"/usr/local/Cellar/tfenv/1.0.2/versions/0.12.16/terraform", "plan"}
2019/11/27 15:47:45 [DEBUG] Attempting to open CLI config file: /Users/jyou/.terraformrc
2019/11/27 15:47:45 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/11/27 15:47:45 [INFO] CLI command args: []string{"plan"}
2019/11/27 15:47:45 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2019/11/27 15:47:45 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2019/11/27 15:47:45 [DEBUG] New state was assigned lineage "9c7a69b6-51f9-4734-d06d-388ed03383ad"
2019/11/27 15:47:45 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2019/11/27 15:47:45 [TRACE] Meta.Backend: instantiated backend of type <nil>
2019/11/27 15:47:45 [DEBUG] checking for provider in "."
2019/11/27 15:47:45 [DEBUG] checking for provider in "/usr/local/Cellar/tfenv/1.0.2/versions/0.12.16"
2019/11/27 15:47:45 [DEBUG] checking for provider in ".terraform/plugins/darwin_amd64"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-aws_v2.26.0_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-external_v1.2.0_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-http_v1.1.1_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-local_v1.3.0_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-null_v2.1.2_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-random_v2.2.1_x4"
2019/11/27 15:47:45 [DEBUG] found provider "terraform-provider-template_v2.1.2_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "local", "1.3.0", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "null", "2.1.2", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "random", "2.2.1", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "template", "2.1.2", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "aws", "2.26.0", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "external", "1.2.0", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4"
2019/11/27 15:47:45 [DEBUG] found valid plugin: "http", "1.1.1", "/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4"
2019/11/27 15:47:45 [DEBUG] checking for provisioner in "."
2019/11/27 15:47:45 [DEBUG] checking for provisioner in "/usr/local/Cellar/tfenv/1.0.2/versions/0.12.16"
2019/11/27 15:47:45 [DEBUG] checking for provisioner in ".terraform/plugins/darwin_amd64"
2019/11/27 15:47:45 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2019/11/27 15:47:45 [INFO] backend/local: starting Plan operation
2019/11/27 15:47:45 [TRACE] backend/local: requesting state manager for workspace "default"
2019/11/27 15:47:45 [TRACE] backend/local: state manager for workspace "default" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2019/11/27 15:47:45 [TRACE] backend/local: requesting state lock for workspace "default"
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: preparing to manage state snapshots at terraform.tfstate
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: existing snapshot has lineage "31bf9dc8-440d-d376-048c-b2c64e88968c" serial 5
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: locking terraform.tfstate using fcntl flock
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: writing lock metadata to .terraform.tfstate.lock.info
2019/11/27 15:47:45 [TRACE] backend/local: reading remote state for workspace "default"
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: reading latest snapshot from terraform.tfstate
2019/11/27 15:47:45 [TRACE] statemgr.Filesystem: read snapshot with lineage "31bf9dc8-440d-d376-048c-b2c64e88968c" serial 5
2019/11/27 15:47:45 [TRACE] backend/local: retrieving local state snapshot for workspace "default"
2019/11/27 15:47:45 [TRACE] backend/local: building context for current working directory
2019/11/27 15:47:45 [TRACE] terraform.NewContext: starting
2019/11/27 15:47:45 [TRACE] terraform.NewContext: resolving provider version selections
2019/11/27 15:47:46 [TRACE] terraform.NewContext: loading provider schemas
2019/11/27 15:47:46 [TRACE] LoadSchemas: retrieving schema for provider type "aws"
2019-11-27T15:47:46.847-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:46.871-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4]
2019-11-27T15:47:46.879-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4 pid=16566
2019-11-27T15:47:46.879-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4
2019-11-27T15:47:46.899-0800 [INFO]  plugin.terraform-provider-aws_v2.26.0_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:46.899-0800
2019-11-27T15:47:46.926-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:46.926-0800 [DEBUG] plugin.terraform-provider-aws_v2.26.0_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin090975709 network=unix timestamp=2019-11-27T15:47:46.925-0800
2019/11/27 15:47:46 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.028-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-aws_v2.26.0_x4 pid=16566
2019-11-27T15:47:47.028-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "external"
2019-11-27T15:47:47.028-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.052-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4]
2019-11-27T15:47:47.056-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4 pid=16567
2019-11-27T15:47:47.056-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4
2019-11-27T15:47:47.065-0800 [INFO]  plugin.terraform-provider-external_v1.2.0_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.064-0800
2019-11-27T15:47:47.093-0800 [DEBUG] plugin.terraform-provider-external_v1.2.0_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin747581978 network=unix timestamp=2019-11-27T15:47:47.093-0800
2019-11-27T15:47:47.093-0800 [DEBUG] plugin: using plugin: version=5
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.149-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-external_v1.2.0_x4 pid=16567
2019-11-27T15:47:47.149-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "http"
2019-11-27T15:47:47.149-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.175-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4]
2019-11-27T15:47:47.178-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4 pid=16568
2019-11-27T15:47:47.178-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4
2019-11-27T15:47:47.188-0800 [INFO]  plugin.terraform-provider-http_v1.1.1_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.188-0800
2019-11-27T15:47:47.219-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:47.219-0800 [DEBUG] plugin.terraform-provider-http_v1.1.1_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin951812655 network=unix timestamp=2019-11-27T15:47:47.218-0800
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.276-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-http_v1.1.1_x4 pid=16568
2019-11-27T15:47:47.276-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "local"
2019-11-27T15:47:47.276-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.299-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4]
2019-11-27T15:47:47.303-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4 pid=16569
2019-11-27T15:47:47.303-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4
2019-11-27T15:47:47.316-0800 [INFO]  plugin.terraform-provider-local_v1.3.0_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.316-0800
2019-11-27T15:47:47.344-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:47.344-0800 [DEBUG] plugin.terraform-provider-local_v1.3.0_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin603895364 network=unix timestamp=2019-11-27T15:47:47.344-0800
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.401-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.3.0_x4 pid=16569
2019-11-27T15:47:47.401-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "null"
2019-11-27T15:47:47.401-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.426-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4]
2019-11-27T15:47:47.430-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4 pid=16570
2019-11-27T15:47:47.431-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4
2019-11-27T15:47:47.440-0800 [INFO]  plugin.terraform-provider-null_v2.1.2_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.440-0800
2019-11-27T15:47:47.467-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:47.468-0800 [DEBUG] plugin.terraform-provider-null_v2.1.2_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin754165409 network=unix timestamp=2019-11-27T15:47:47.467-0800
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.522-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-null_v2.1.2_x4 pid=16570
2019-11-27T15:47:47.523-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "random"
2019-11-27T15:47:47.523-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.548-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4]
2019-11-27T15:47:47.551-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4 pid=16571
2019-11-27T15:47:47.552-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4
2019-11-27T15:47:47.563-0800 [INFO]  plugin.terraform-provider-random_v2.2.1_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.563-0800
2019-11-27T15:47:47.590-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:47.590-0800 [DEBUG] plugin.terraform-provider-random_v2.2.1_x4: plugin address: network=unix address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin868527942 timestamp=2019-11-27T15:47:47.590-0800
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.647-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.2.1_x4 pid=16571
2019-11-27T15:47:47.647-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] LoadSchemas: retrieving schema for provider type "template"
2019-11-27T15:47:47.647-0800 [INFO]  plugin: configuring client automatic mTLS
2019-11-27T15:47:47.672-0800 [DEBUG] plugin: starting plugin: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4 args=[/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4]
2019-11-27T15:47:47.675-0800 [DEBUG] plugin: plugin started: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4 pid=16572
2019-11-27T15:47:47.675-0800 [DEBUG] plugin: waiting for RPC address: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4
2019-11-27T15:47:47.685-0800 [INFO]  plugin.terraform-provider-template_v2.1.2_x4: configuring server automatic mTLS: timestamp=2019-11-27T15:47:47.685-0800
2019-11-27T15:47:47.713-0800 [DEBUG] plugin: using plugin: version=5
2019-11-27T15:47:47.713-0800 [DEBUG] plugin.terraform-provider-template_v2.1.2_x4: plugin address: address=/var/folders/g0/2sykh8h57knch4c5ybs5qqrnw0b7t_/T/plugin795925283 network=unix timestamp=2019-11-27T15:47:47.713-0800
2019/11/27 15:47:47 [TRACE] GRPCProvider: GetSchema
2019/11/27 15:47:47 [TRACE] GRPCProvider: Close
2019-11-27T15:47:47.772-0800 [DEBUG] plugin: plugin process exited: path=/Users/jyou/Desktop/tmp/iam/.terraform/plugins/darwin_amd64/terraform-provider-template_v2.1.2_x4 pid=16572
2019-11-27T15:47:47.772-0800 [DEBUG] plugin: plugin exited
2019/11/27 15:47:47 [TRACE] terraform.NewContext: complete
2019/11/27 15:47:47 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2019/11/27 15:47:47 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
panic: inconsistent map element types (cty.Object(map[string]cty.Type{"action":cty.Tuple([]cty.Type{cty.String, cty.String}), "condition":cty.EmptyObject, "effect":cty.String, "resources":cty.Tuple([]cty.Type{cty.String, cty.String})}) then cty.Object(map[string]cty.Type{"action":cty.Tuple([]cty.Type{cty.String}), "condition":cty.EmptyObject, "effect":cty.String, "resources":cty.Tuple([]cty.Type{cty.String})}))

goroutine 58 [running]:
github.com/zclconf/go-cty/cty.MapVal(0xc0000bb260, 0xc0000bb200, 0xc000040e55, 0x5, 0xc0000bb338)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/value_init.go:207 +0x538
github.com/zclconf/go-cty/cty/convert.conversionObjectToMap.func2(0x2f0edc0, 0xc00000e198, 0x270f5a0, 0xc00059dc50, 0xc00079ad00, 0x1, 0x1, 0x45b76d0, 0x0, 0x100bf98, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/convert/conversion_collection.go:338 +0x6c3
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2f0edc0, 0xc00000e198, 0x270f5a0, 0xc00059dc50, 0x0, 0x0, 0x0, 0xc0000cd420, 0xc00079ace0, 0x2f0ed80, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/convert/conversion.go:34 +0xef
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x2f0edc0, 0xc00000e198, 0x270f5a0, 0xc00059dc50, 0xc00079ace0, 0xc0001e7bf0, 0x403d3d0, 0x0, 0xc0004456c8, 0x100e3d9)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/convert/conversion.go:141 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x2f0edc0, 0xc00000e198, 0x270f5a0, 0xc00059dc50, 0x2f0ed80, 0xc00007cab0, 0xc0000bb5e8, 0x1000000010495de, 0x1056300, 0xc0000bb5e7, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/convert/public.go:51 +0x1af
github.com/hashicorp/terraform/terraform.checkInputVariables(0xc0001e7bf0, 0xc00059ddd0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/variables.go:255 +0x191
github.com/hashicorp/terraform/terraform.NewContext(0xc000410dc0, 0x0, 0x0, 0xc0002cda00, 0x1)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/context.go:214 +0x90f
github.com/hashicorp/terraform/backend/local.(*Local).contextDirect(0xc0001cb860, 0xc00032f880, 0x0, 0x0, 0xc0005a2a58, 0x0, 0x0, 0x0, 0x0, 0xc00015dfc0, ...)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_local.go:161 +0x308
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc0001cb860, 0xc00032f880, 0x1, 0x1, 0x2, 0xc00016ab60, 0x1, 0x2, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_local.go:94 +0xd27
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc0001cb860, 0x2f0d000, 0xc0000bfe40, 0x2f0d000, 0xc0000bfe80, 0xc00032f880, 0xc0000bfe00)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:70 +0x2a3
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc00007c100, 0xc00007c110, 0xc00007c120, 0xc00032f880, 0xc0001cb860, 0xc0000bfe00, 0xc00007c0f0, 0x2f0d000, 0xc0000bfe40, 0x2f0d000, ...)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:356 +0x135
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:340 +0x36b
```

**Comment by jwyou1991** [comment]
Tried with map data structure. same result 
```
variable "custom"{
    type = map(object({
        action = any
        resources = any
        effect = string
        condition = any
    }))
}


 custom = {      
        "rule1" = {
                action = [
                        "s3:Get*",
                        "s3:List*"
                ]
                resources = [
                        "arn:aws:s3:::awsbilling-master-groupon-com",
                        "arn:aws:s3:::mybucket1-test"
                ]
                effect = "Allow"
                condition = {}
        }
        "rule2" = {
                action = ["s3:CreateBucket"]
                resources = ["arn:aws:s3:::mybucket-test"]
                effect = "Allow"
                condition = {}
        }
}
```

**Comment by jwyou1991** [comment]
Hi, @danieldreier any updates? Thanks.

**Comment by pselle** [comment]
@jwyou1991  You can be more specific in your type definitions -- when you got this warning:

> Additional note on the variable setup, originally I would like to set the object.action & object.resource to be list type, but terraform would complain about invalid type specification

It was because if you have a list, you must specify the types contained in the list.

```
    type = list(object({
        action = list(string)
        resource = list(string)
        affect = string
# note, I do not know what you want to be in condition, so I just made it empty object, 
# but you'd need to define a type definition that works for you
        condition = object({})
    }))
```

The panic is not ideal, but given the `any` type, the type system is doing its [current] best to interpret what you've put in, and unfortunately, for one of the values, it's deemed that you meant "a tuple of two things" and "a tuple of one thing" when you actually want a list. The type definition above should hopefully help you get going!

**Comment by pselle** [comment]
For future readers, pulling the panic out:


panic: inconsistent map element types (
    
cty.Object(map[string]cty.Type{"action":cty.Tuple([]cty.Type{cty.String, cty.String}), "condition":cty.EmptyObject, "effect":cty.String, **"resources":cty.Tuple([]cty.Type{cty.String, cty.String})**}) then 

cty.Object(map[string]cty.Type{"action":cty.Tuple([]cty.Type{cty.String}), "condition":cty.EmptyObject, "effect":cty.String, **"resources":cty.Tuple([]cty.Type{cty.String})**}))

**Comment by mildwonkey** [comment]
Hello! This has since been fixed and is working as expected in master. The fix will be included in the 0.13 release. 
Thank you again for reporting it!
